docs/specification/new-spec/consensus-reactor.md

```diff
@@ -0,0 +1,239 @@
+for decoding message received from a peer and for adequate processing of the message depending on its type and content.
+The processing normally consists of updating the known peer state and for some messages 
+(`ProposalMessage`, `BlockPartMessage` and `VoteMessage`) also forwarding message to ConsensusState module 
+for further processing. In the following text we specify the core functionality of those separate unit of executions 
+that are part of the Consensus Reactor. 
+
+## Consensus State
+
+Consensus State handles execution of the Tendermint BFT consensus algorithm. It processes votes and proposals, 
+and upon reaching agreement, commits blocks to the chain and executes them against the application.
+The internal state machine receives input from peers, the internal validator and from a timer.
+
+Inside Consensus State we have the following units of execution: TimeoutTicker and Receive Routine.
+TimeoutTicker is a timer that schedules timeouts conditional on the height/round/step that are processed 
+by the Receive Routine. 
+
+
+### Receive Routine
+
+Receive Routine of the ConsensusState handles messages which may cause internal consensus state transitions.
+It is the only routine that updates RoundState that contains internal consensus state. 
+Updates (state transitions) happen on timeouts, complete proposals, and 2/3 majorities. 
+It receives messages from peers, internal validators and from TimeoutTicker 
+and invokes the corresponding handlers, potentially updating the RoundState. 
+The details of the protocol (together with formal proofs of correctness) implemented by the Receive Routine are 
+discussed in separate document (see [spec](https://github.com/tendermint/spec)). For understanding of this document
+it is sufficient to understand that the Receive Routine manages and updates RoundState data structure that is 
+then extensively used by the gossip routines to determine what information should be sent to peer processes.
+
+## Round State
+
+RoundState defines the internal consensus state. It contains height, round, round step, a current validator set,
+a proposal and proposal block for the current round, locked round and block (if some block is being locked), set of 
+received votes and last commit and last validators set.  
+
+```
+type RoundState struct {
+	Height             int64 
+	Round              int
+	Step               RoundStepType
+	Validators         ValidatorSet
+	Proposal           Proposal
+	ProposalBlock      Block
+	ProposalBlockParts PartSet
+	LockedRound        int
+	LockedBlock        Block
+	LockedBlockParts   PartSet
+	Votes              HeightVoteSet
+	LastCommit         VoteSet 
+	LastValidators     ValidatorSet
+}    
+``` 
+
+Internally, consensus will run as a state machine with the following states:
+RoundStepNewHeight, RoundStepNewRound, RoundStepPropose, RoundStepPropose, RoundStepPrevote,       
```

Comment:
RoundStepProposeWait

Reply:
Looks like `RoundstepPropose` is twice in the list.

Reply:
Yep, it should be RoundStepProposeWait second time.